Autodocking

tul1 opened the pull request:
Autodocking routine: when the battery of Leonardo is running low, Leonardo have to go docking.

julianmateu commented on src/laptop_battery_monitor/scripts/laptop_battery.py:
```diff
@@ -249,6 +249,8 @@ def __init__(self):
 
         # Battery info
         self._batt_acpi_path = rospy.get_param('~acpi_path', "/sys/class/power_supply/BAT0")
+        # self._batt_acpi_path = rospy.get_param('~acpi_path', "/home/tul1/Desktop/MOCK_BAT")
```
Please remove, this file should not be changed.

tul1 replied:
done!

tul1 commented:
@julianmateu the code is ready for a review.

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+        elif status == GoalStatus.RECALLED  : state='RECALLED'
+        elif status == GoalStatus.LOST      : state='LOST'
+        # Print state of action server
+        print 'Result - [ActionServer: ' + state + ']: ' + result.text
```
ROS provides a `rosout` topic for [logging information from nodes](http://wiki.ros.org/rospy/Overview/Logging). I think that is the preferred method, right @ernestmc?

ernestmc replied:
Yes, you can use
```
rospy.logdebug("Whatever you want to log")
```

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+        self._go_docking = False
+
+    def _doneCb(self, status, result):
+        if status == GoalStatus.PENDING   : state='PENDING'
```
Style: alignment is inconsistent in these lines.

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+        rospy.on_shutdown(self._client.cancel_goal)
+        self._go_docking = False
+
+    
```
Extra line

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+        rospy.spin()
+
+
+
```
Extra line. See [PEP-8](https://www.python.org/dev/peps/pep-0008/#blank-lines):

> Surround top-level function and class definitions with two blank lines.
> Method definitions inside a class are surrounded by a single blank line.

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+        self._ros_node = rospy.init_node('dock_drive_client_py', anonymous=True)
+        self._client = actionlib.SimpleActionClient('dock_drive_action', AutoDockingAction)
+        self._diagnostic_agg_sub = rospy.Subscriber("/diagnostics_agg", DiagnosticArray, self._listen_batteries)
+        self._go_docking = False
```
For a first version this might be OK, but I would suggest to use a more descriptive name, and make sure the behavior is the intended one. With the current implementation the robot will go docking only once and never dock again.

For example, here this variable will be set to true when the docking action has started, so a more descriptive name would be `_docking_action_started` or `_docking_in_progress`. In that case, you would need to make sure that it is set to false again once the action finishes. But you would also need another flag that tells you if the robot is already docked and charging, so that you don't trigger the auto docking in the case it is charging (you could read the battery state for that).

So the logic will be something like:
If not charging and not docking_in_progress then dock (if the battery level is low).

tul1 replied:
I forgot to change the state of `_go_docking` when is called `_doneCb`. Changing that state there I think the Leonardo is going to behave as expected.

julianmateu commented on src/laptop_battery_monitor/scripts/DockDriveActionClient.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class AutoDocking:
+    BATTERY_THRESHOLD = 4
```
This is OK for now, but it would be good to make it a parameter configurable from the launch file.

tul1 commented:
@ernestmc I'd added the autodocking routine to a package. Would you give a look of what I've done?

tul1 commented:
@ernestmc I could be able to build the whole project with `catkin_make` as you can see in the screenshot beneath:

![image](https://user-images.githubusercontent.com/10169300/43418079-4559fea8-9413-11e8-9cde-7b6e6b01a593.png)

However, I cannot run the `auto_docking` with `rosrun`, it doesn't even show this package as an option when I press `tab`. Running `rosrun auto_docking` I get the following error:

```
Usage: rosrun [--prefix cmd] [--debug] PACKAGE EXECUTABLE [ARGS]
  rosrun will locate PACKAGE and try to find
  an executable named EXECUTABLE in the PACKAGE tree.
  If it finds it, it will run it with ARGS.
```
So the package isn't in the namespace.

I pushed the code so you can have a look @ernestmc . Let me know if you find anything wrong that could cause this problem. Thanks :-)

ernestmc commented on src/auto_docking/package.xml:
```diff
@@ -0,0 +1,27 @@
+<package>
+  <name>auto_docking</name>
```
The name of the package seems too generic. I think what we are doing is monitoring the battery state and executing actions, maybe it can be something like `leonardo_battery_guard` or something similar.

ernestmc commented on src/auto_docking/package.xml:
```diff
@@ -0,0 +1,27 @@
+  <version>0.1.103</version>
+
+  <description>
+    The auto_docking package
```
Please add a better description of what the package does.

ernestmc commented on src/auto_docking/package.xml:
```diff
@@ -0,0 +1,27 @@
+  <buildtool_depend>catkin</buildtool_depend>
+  <build_depend>kobuki_msgs</build_depend>
+  <build_depend>actionlib_msgs</build_depend>
+  <build_depend>diagnostic_msgs</build_depend>
```
Since this is a python package and nothing gets compiled, the `build_depends` are not really needed.

tul1 replied:
I don't really know why but apparently these build deps are necessary to build the package. The laptop monitor package also includes them.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
```
Again, a better and more descriptive name other than `auto_docking_client` which is too generic.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/env python
+
+import roslib; roslib.load_manifest('kobuki_auto_docking')
```
This is not needed. Please remove the line. (Also why are you using `;` this is not C++!!)

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+from diagnostic_msgs.msg import DiagnosticArray
+
+
+class AutoDocking:
```
Class should inherit from Object.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+    BATTERY_THRESHOLD = 4
+
+    def __init__(self):
+        self._ros_node = rospy.init_node('dock_drive_client_py', anonymous=True)
```
Remove the `py` part. The language in which the node is written is irrelevant to the ros system. Use something that describes what it does.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        self._diagnostic_agg_sub = rospy.Subscriber("/diagnostics_agg", DiagnosticArray, self._listen_batteries)
+        self._go_docking = False
+
+    def _doneCb(self, status, result):
```
Add docstring to the method.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        elif status == GoalStatus.PREEMPTING: state='PREEMPTING'
+        elif status == GoalStatus.RECALLING: state='RECALLING'
+        elif status == GoalStatus.RECALLED: state='RECALLED'
+        elif status == GoalStatus.LOST: state='LOST'
```
What do we need all this logic for?

tul1 replied:
It's something that has gotten left from the kobuki autodocking routine. It just transcripts the `GoalStatus`. I left it because it might help to debug later.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        print 'Result - [ActionServer: ' + state + ']: ' + result.text
+        self._go_docking = False
+
+    def _feedbackCb(self, feedback):
```
Do we need the feedback? Otherwise remove.

tul1 replied:
just for debugging. I would keep it until we are 100% sure that everything is working as expected.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+            return        
+        goal = AutoDockingGoal()
+        self._client.send_goal(goal, done_cb=self._doneCb, feedback_cb=self._feedbackCb)
+        print 'Goal: Sent.'
```
Prefer ros logging.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        self._go_docking = False
+    
+    def _listen_batteries(self, data):
+        batteries_names = ['/Power System/Laptop Battery', "/Power System/Battery"]
```
Always use double quotes.

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        batteries_names = ['/Power System/Laptop Battery', "/Power System/Battery"]
+        batteries_values = [element.values for element in data.status if element.name in batteries_names]
+        laptop_battery_percentage = float(filter(lambda x: x.key == "Percentage (%)", batteries_values[0])[0].value)
+        kuboki_battery_percentage = float(filter(lambda x: x.key == "Percent", batteries_values[1])[0].value)
```
`kuboki` --> `kobuki`

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+
+
+class AutoDocking:
+    BATTERY_THRESHOLD = 4
```
Use [ros parameters](http://wiki.ros.org/rospy/Overview/Parameter%20Server) for these values.

julianmateu commented:
@tul1 did you source the `setup.sh` file?

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        
+        if self._go_docking is False:
+            if kuboki_battery_percentage < self.BATTERY_THRESHOLD or laptop_battery_percentage < self.BATTERY_THRESHOLD:
+                print kuboki_battery_percentage
```
Remove debugging code or use ros logging

tul1 commented:
Hey @julianmateu ! Yes, I did. Maybe the problem wasn't the package but the python script. I'm getting the following error when running: `rosrun auto_docking auto_docking_client.py`
```
[rosrun] Couldn't find executable named auto_docking_client.py below /home/tul1/Desktop/proyectos/leonardo_turtlebot/src/auto_docking
[rosrun] Found the following, but they're either not files,
[rosrun] or not executable:
[rosrun]   /home/tul1/Desktop/proyectos/leonardo_turtlebot/src/auto_docking/scripts/auto_docking_client.py
```
(the autocomplete `tab` just works until `auto_docking`)

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+        if self._go_docking is False:
+            if kuboki_battery_percentage < self.BATTERY_THRESHOLD or laptop_battery_percentage < self.BATTERY_THRESHOLD:
+                print kuboki_battery_percentage
+                self._dock_drive_client()
```
It seems reasonable that this method is called `do_docking` or `go_dock` since it invokes the action, doesn't it?

tul1 replied:
It does!

ernestmc commented on src/auto_docking/scripts/auto_docking_client.py:
```diff
@@ -0,0 +1,69 @@
+            if kuboki_battery_percentage < self.BATTERY_THRESHOLD or laptop_battery_percentage < self.BATTERY_THRESHOLD:
+                print kuboki_battery_percentage
+                self._dock_drive_client()
+                self._go_docking = True
```
I understand correctly this flag is actually telling if the robot is in auto docking mode, right? Shouldn't it then be called something like `docking`, `doing_docking`, `docking_invoked` or something like that?

tul1 replied:
I like `doing_docking`. I'm going to change it.

tul1 commented:
@julianmateu We solved the problem with @ernestmc just changing permissions of the python script auto_docking_client.py to be executable.

tul1 commented:
@ernestmc trying to set argument from the launch file I've gotten the following error:
```
ERROR: cannot launch node of type [leonardo_battery_guard/dock_drive_client_py]: can't locate node [dock_drive_client_py] in package [leonardo_battery_guard]
```
would you have any idea where is this coming from?
I've just pushed the launch file.

tul1 commented:
@ernestmc @julianmateu I'm ready for a new review!

julianmateu commented on src/leonardo_battery_guard/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+  PROGRAMS
+    scripts/battery_guard.py
+    DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
```
Missing new line at the end of the file.

julianmateu commented on src/leonardo_battery_guard/launch/leonardo_battery_guard.launch:
```diff
@@ -0,0 +1,6 @@
+
```
Extra line.

julianmateu commented on src/leonardo_battery_guard/scripts/battery_guard.py:
```diff
@@ -0,0 +1,79 @@
+        if self._doing_docking is False:
+            if kobuki_battery_percentage < self.BATTERY_THRESHOLD or laptop_battery_percentage < self.BATTERY_THRESHOLD:
+                self._go_dock()
+                self._doing_docking = True
```
What happens here if the robot is docked and charging but the battery level is below the threshold? Does the docking action get triggered? Have you tested this case?

tul1 replied:
Good point @julianmateu ! I added some logic to prevent that case.

tul1 commented:
@julianmateu @ernestmc I added a navigation action client to the node to send Leonardo close to the dock station before it starts autodocking.
I tried the code in Leonardo and in the turtlebot gazebo simulation but I could make it work and no error were printed. Tomorrow I'll continue testing different alternatives but I think that by this point the code I pushed is ready for a new review.
I added some TODO comments in the code, please let me know what do you think about them.

tul1 commented:
@julianmateu @ernestmc the code is ready for a new review. 
It's working and it has been tested satisfactory twice on Leonardo. It's going to need further field tests to adjust details.

tul1 commented:
@julianmateu FYI with @ernestmc  we handled to correct the last code errors and we were able to autodock Leonardo.

tul1 commented:
This code here is ready to be reviewed, it's "working" and it has been tested several times in Leonardo. I quoted "working" because the code or system is going to need further adjustments. For instance, the biggest issue we have lies in the autodocking routine. In fact, Leonardo rarely docks, it just approaches the dock and hits the base.

Another possible issue is that we think that the routine is being triggered in condition where it shouldn't. For example, while the turtlebot is charging. We couldn't verify this but it happened twice that Leonardo was running the autodocking routine after being docked.

Also we talked offline with @ernestmc about some changes that occurred from master to dev, and before merging this branch the file system should be reorganized.

Last comment, in order to run the routine for debugging proposes I added a service named `/run_autodocking` and to call it the following line should be invoked in the terminal: `rosservice call /run_autodocking "{}"`
